src/lsm/block_free_set.zig

```diff
@@ -1,37 +1,421 @@
+        assert(shard_size == 4096);
+        assert(@bitSizeOf(MaskInt) == 64);
+    }
+
+    pub fn init(allocator: *mem.Allocator, blocks_count: usize) !BlockFreeSet {
+        assert(shard_size <= blocks_count);
+        assert(blocks_count % shard_size == 0);
+        assert(blocks_count % @bitSizeOf(usize) == 0);
+
+        // Round up to ensure that every block bit is covered by the index.
+        const shards_count = div_ceil(blocks_count, shard_size);
+        var index = try DynamicBitSetUnmanaged.initFull(shards_count, allocator);
+        errdefer index.deinit(allocator);
+
+        var blocks = try DynamicBitSetUnmanaged.initFull(blocks_count, allocator);
+        errdefer blocks.deinit(allocator);
 
-    pub fn init(allocator: *mem.Allocator, count: usize) !BlockFreeSet {
         return BlockFreeSet{
-            .free = try std.bit_set.DynamicBitSetUnmanaged.initFull(count, allocator),
+            .index = index,
+            .blocks = blocks,
+            .recent = .{},
         };
     }
 
     pub fn deinit(set: *BlockFreeSet, allocator: *mem.Allocator) void {
-        set.free.deinit(allocator);
+        set.index.deinit(allocator);
+        set.blocks.deinit(allocator);
     }
 
-    // TODO consider "caching" the first set bit to speed up subsequent acquire() calls
-    pub fn acquire(set: *BlockFreeSet) u64 {
-        // TODO: To ensure this "unreachable" is never reached, the leader must reject
-        // new requests when storage space is too low to fulfill them.
-        const bit = set.free.findFirstSet() orelse unreachable;
-        set.free.unset(bit);
-        const address = bit + 1;
+    /// Marks a free block as allocated, and returns the address. Panics if no blocks are available.
+    pub fn acquire(set: *BlockFreeSet) ?u64 {
+        const block = blk: {
+            if (set.recent.pop()) |block| {
+                break :blk block;
+            } else if (set.index.findFirstSet()) |shard| {
+                break :blk set.find_free_block_in_shard(shard) orelse unreachable;
+            } else return null;
+        };
+        const shard = block / shard_size;
+        assert(set.blocks.isSet(block));
+        assert(set.index.isSet(shard));
+
+        set.blocks.unset(block);
+        // Update the index when every block in the shard is allocated.
+        if (set.find_free_block_in_shard(shard) == null) set.index.unset(shard);
+
+        const address = block + 1;
         return @intCast(u64, address);
     }
 
+    fn find_free_block_in_shard(set: *BlockFreeSet, shard: usize) ?usize {
+        const shard_start = shard * shard_size;
+        const shard_end = shard_start + shard_size;
+        assert(shard_start < set.blocks.bit_length);
+
+        return find_first_set_bit(set.blocks, shard_start, shard_end);
+    }
+
+    fn is_free(set: *BlockFreeSet, address: u64) bool {
+        const block = address - 1;
+        return set.blocks.isSet(block);
+    }
+
+    /// Marks the specified block as free.
```

Comment:
```suggestion
    /// Given the address, marks an allocated block as free.
```

Just mirroring the doc comment so that it's symmetrical with that of `acquire()`.